diff-hl.el

```diff
@@ -1525,42 +1534,119 @@ view buffer.
-      (when diff-hl-mode
-        (when (bound-and-true-p diff-hl-amend-mode)
-          (diff-hl-amend-mode -1))
-        (diff-hl-update)))))
+  (unless current-prefix-arg
+    (message "Set global reference revision to %s" rev))
+  (cond ((equal '(4) current-prefix-arg)
+         (diff-hl-reset-reference-rev))
+        ((equal '(16) current-prefix-arg)
+         (diff-hl-reset-reference-rev '(4)))
+        (t (setq-default diff-hl-reference-revision rev)
+           (dolist (buf (buffer-list))
+             (with-current-buffer buf
+               (when (and diff-hl-mode
+                          (not (local-variable-p 'diff-hl-reference-revision)))
+                 (when (bound-and-true-p diff-hl-amend-mode)
+                   (diff-hl-amend-mode -1))
+                 (diff-hl-update)))))))
 
 ;;;###autoload
-(defun diff-hl-reset-reference-rev ()
-  "Reset the reference revision globally to the most recent one."
-  (interactive)
-  (setq diff-hl-reference-revision nil)
+(defun diff-hl-set-reference-rev-in-project (rev)
+  "Set the reference revision in the current project to REV.
+When called interactively, REV read with completion.
+
+When called with a prefix argument, reset to the most recent one instead.
+
+The default value chosen using one of methods below:
+
+- In a log view buffer, it uses the revision of current entry.
+Call `vc-print-log' or `vc-print-root-log' first to open a log
+view buffer.
+- In a VC annotate buffer, it uses the revision of current line.
+- In other situations, it uses the symbol at point.
+
+Projects whose reference was set with this command are unaffected by
+subsequent changes to the global reference (see
+`diff-hl-set-reference-rev`).
+
+Also notice that this will disable `diff-hl-amend-mode' in
+buffers that enables it, since `diff-hl-amend-mode' overrides its
+effect."
+  (interactive
+   (if current-prefix-arg current-prefix-arg
+     (let* ((def (or (and (equal major-mode 'vc-annotate-mode)
+                          (car (vc-annotate-extract-revision-at-line)))
+                     (log-view-current-tag)
+                     (thing-at-point 'symbol t)))
+            (prompt (if def
+                        (format "Reference revision (default %s): " def)
+                      "Reference revision: ")))
+       (list (vc-read-revision prompt nil nil def)))))
+  (unless rev
+    (user-error "No reference revision specified"))
+  (let ((name (project-name (project-current))))
+    (cond
+     ;; reset
+     (current-prefix-arg
+      (diff-hl-set-reference-rev-in-project-internal nil)
+      (message "Reference revision reset to the most recent revision (project %s)" name))
+     ;; set
+     (t (diff-hl-set-reference-rev-in-project-internal rev)
+        (message "Showing changes against %s (project %s)" rev name)))))
+
+(defun diff-hl-set-reference-rev-in-project-internal (rev)
+  (let* ((proj (project-current))
+         (class (intern (concat "diff-hl-" (project-name proj)))))
```

Comment:
project-name can easily conflict, you probably should use (some variation on) project-root instead?

Reply:
Thanks, indeed.
Switched to `project-root` directly and made sure that symbols with ~,/ and spaces can be interned.
Symbol doesn't look fancy (e.g. `diff-hl-~/git/diff-hl/`), but it's never exposed to user and works.